src/apply.jl

```diff
@@ -170,7 +170,7 @@ end
 
 function ITensors.apply(
   o::ITensor,
-  ψ::ITensorNetwork;
+  ψ::Union{ITensorNetwork,TreeTensorNetwork};
```

Comment:
Can we use `AbstractITensorNetwork` here? My understanding is that if the network is `VidalITensorNetwork` it will call the other version, otherwise it will call this version (i.e. the Julia compiler will dispatch to the most specialized version it can find).

Reply:
I was thinking this too but when I did this I got the following bug from `test/test_apply.jl` which is trying to use `apply` on an `VidalITensorNetwork`
<img width="986" alt="image" src="https://github.com/mtfishman/ITensorNetworks.jl/assets/51231103/f7b00ba7-f731-4f37-9d42-8b213ef3025c">

Suggesting that Julia doesn't see it as a more specialized version, which I am surprised by?

Reply:
I think that issue is caused by the operator being constrained to `o::ITensor` in one and `o::Union{ITensor,NamedEdge}` in the other. That may be confusing the compiler, since it can't decide if it should choose the version of the method based on the type of `o` or `ψ`.

What if we just remove the type constraint on `o` in both?

Reply:
Yeah good spot, that's why it's happening.

Okay I have removed the constraint.